documentation/Investigating-FIR-as-an-MLIR-dialect.md

```diff
@@ -0,0 +1,404 @@
+be built and added through the extensibility features of MLIR. This
+process of extending MLIR is called adding a dialect. In our case, we'll
+add a FIR dialect for representation of Fortran programs.
+
+Specifically, this means that we can proceed with the plan as laid out
+in the Fortran IR design document with the advantages of not having to
+build from scratch our own
+
+-    IR infrastructure    
+-    LLVM IR bridge  
+-    Common optimization passes (For example: CSE, DCE, loop fusion, loop
+     invariant code motion, loop tiling, loop unroll and jam, loop
+     unrolling, vectorization, etc.)
+    
+A dialect can be thought of
+as its own distinct IR, where the design and
+semantics of the IR are domain-specific for the purposes of the
+dialect. Specifically, MLIR supports a modern IR design with
+
+-    A predefined set of operations with precise semantics
+-    Structural/logical grouping of sets of operations
+-    Explicit control flow
+-    Explicit data-flow
+-    Strong typing
+
+-    Meta information
+
+     Informally, one can construct MLIR operations that are
+     conceptually composable containers. These containers can have
+     strongly typed metadata attached to them that can describe
+     any properties of interest. For example, one can define an
+     operation called "execute" and define it to have an attribute
+     "order", which may have a value that describes a sequential
+     execution order, a fully parallel unordered execution, or some
+     valid execution partial ordering. What meta information is
+     interesting depends on the context and is up to the MLIR user.
+
+This document describes FIR, which is a set of extensions upon the
+standard MLIR dialect. When lowering a Fortran parse tree to FIR,
+the compiler will produce a mix of operations defined in the FIR
+dialect, the MLIR __standard dialect__, and other pre-defined
+dialects, such as the [affine dialect](https://bit.ly/2XYDPNj).
+
+#### Disclaimers
+
+This document is not a tutorial on compilers, Fortran, MLIR, LLVM,
+intermediate representations, abstraction levels, semantics, nor type
+theory.
+
+## Requirements Unchanged
+
+The FIR dialect has the same requirements as spelled out in the
+Fortran IR design document. Specifically, the FIR dialect will
+capture the control flow structure of the Fortran source-level
+program. At the highest level of abstraction, Fortran computations
+will be captured as coarse-grain opaque
+operations corresponding to Fortran expressions
+(with links to `Fortran::evaluate::Expr<T>` objects from the
+front-end), where only the peripheral use-def information is
+exposed. These operations will, of course, necessarily be lowered
+to other various MLIR dialects, where they can be
+optimized.
+The compiler will continue to lower the representation
+of the input with successive transformations
+to MLIR's LLVM IR dialect, the target independent LLVM IR, machine
```

Comment:
At the FIR/MLIR layer will everything be finally converted to the LLVM IR dialect before LLVM IR is generated?
Also, will FIR/MLIR have all information needed and hence it won't be necessary to consult the parse tree for LLVM IR generation?